source/common/common/log_state_utils.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <sstream>
+
+namespace Envoy {
+
+#define LOG_MEMBER(member) ", " #member ": " << (member)
+
+#define LOG_OPTIONAL_MEMBER(member)                                                                \
+  ", " #member ": " << (member.has_value() ? absl::StrCat(member.value()) : "null")
+
+// Macro assumes local member variables
+// os (ostream)
+// indent_level (int)
+#define LOG_DETAILS(member)                                                                        \
+  do {                                                                                             \
+    os << spaces << #member ": ";                                                                  \
+    if (member != nullptr) {                                                                       \
+      os << "\n";                                                                                  \
+      (member)->logState(os, indent_level + 1);                                                    \
+    } else {                                                                                       \
+      os << spaces << "null\n";                                                                    \
+    }                                                                                              \
+  } while (false)
+
+// Return the const char* equivalent of string(level*2, ' '), without dealing
+// with string creation overhead. Cap arbitrarily at 6 as we're (hopefully)
+// not going to have nested objects deeper than that.
+inline const char* spacesForLevel(int level) {
```

Comment:
Q: would it be easier and also fairly readable to build a proto with this information and have the text proto dump do the indenting etc?

Reply:
Possibly, but internally we've found dumping session state in debug logs (error/warning) also useful so I'd like to minimize string copies where possible.

Reply:
Not sure I fully grok what is to be gained performance wise by minimizing string copies here.

Reply:
in-house we often log session state in the critical path. We also dump all active sessions on crash.  Doing lots of allocation and deallocation is a bad idea in crash handlers, as it increases risks of subsequent failure.
Meanwhile tracking proto structures for all dumped session state is a ton of busy work, which doesn't seem worthwhile given the gain of removing a small utility function

Reply:
Yeah, I find it a bit surprising we would be logging anything significant here on the critical path, which is why I was asking. Can you elaborate on this?

I'm not wedded to proto here either. What you have is fine, I'd just like to make sure we're not planning on doing ASCII art on some critical path.